src/store/Controller.cc

```diff
@@ -354,19 +379,40 @@ Store::Controller::accumulateMore(StoreEntry &entry) const
 void
 Store::Controller::markForUnlink(StoreEntry &e)
 {
-    if (transients && e.mem_obj && e.mem_obj->xitTable.index >= 0)
+    if (transients)
         transients->markForUnlink(e);
-    if (memStore && e.mem_obj && e.mem_obj->memCache.index >= 0)
+    if (memStore)
         memStore->markForUnlink(e);
-    if (swapDir && e.swap_filen >= 0)
+    if (swapDir)
         swapDir->markForUnlink(e);
 }
 
 void
+Store::Controller::unlinkByKeyIfFound(const cache_key *key)
+{
+    if (StoreEntry *entry = static_cast<StoreEntry*>(hash_lookup(store_table, key))) {
+        debugs(20, 3, "got in-transit entry: " << *entry);
+        entry->release(true);
+        // The entry should be locked (but see find()). Return here because the
+        // release of a locked entry should take care of all its stores.
```

Comment:
> However, the "all of its stores" logic implies that release() may ignore Store entries that are not attached to the entry object.

Looks like this comment is not precise: release() should mark not attached entry for deletion from storages via Controller::markForUnlink(), and corresponding markForUnlink() storage methods.

Reply:
BTW, Controller::unlinkByKeyIfFound() is only called from purgeEntriesByUrl() with precondition that Store::Root().get() returned nil. That makes me thinking that the "if" block in Controller::unlinkByKeyIfFound() is not necessary.

Reply:
> Code: `release(true); // takes cares of all its stores if any`

> Alex: the "all of its stores" logic implies that release() may ignore Store entries that are not attached to the entry object

> Eduard: release() should mark not attached entry for deletion from storages via Controller::markForUnlink

Whether release() _should_ unlink unattached entries is actually debatable, but, for this discussion, let's assume that release() does unlink unattached entries.

> Eduard: Looks like [Alex] comment is not precise

My comment seems OK, but the source code comment implies something that it should not imply. In other words, if release() unlinks unattached entries, then we need to fix the comment. Something like this may work better:

```
// unlink all Store entries with entry->key, even unattached ones
entry->release(true);
return;
```

And to minimize similar doubts in the future, we should improve the release() description as well. For example, something like this would be better than the current text IMO (we are running rather thin on Store-related terminology, but let's not try to fix that here, in a method description):

```
/// Removes all unlocked (and marks for eventual removal all locked) Store
/// entries, including attached and unattached entries that have our key.
/// Also destroys us if we are unlocked or makes us private otherwise.
virtual void release(const bool shareable = false);
```

Reply:
> Controller::unlinkByKeyIfFound() is only called from purgeEntriesByUrl() with precondition that Store::Root().get() returned nil. That makes me thinking that the "if" block in Controller::unlinkByKeyIfFound() is not necessary.

AFAICT, that thinking does not match the code because Root().get() can return nil when `store_table` has a matching entry. See the markedForDeletion() check in Controller::find().

And, more importantly, we should not make nil get() a hidden precondition for calling Root().unlinkByKeyIfFound() because some caller will eventually forget to check that. Your comment and several similar earlier discussions all suggest to me that we need to revise the unlink/markForUnlink API. I hope we can do that after this PR. For example, it seems likely that a chunk of the current purgeEntriesByUrl() logic should go directly inside Controller::unlinkByKeyIfFound() instead of appearing as preconditions for (or alternative actions to) that method call.

Reply:
Done: fixed comments, the code is the same.